### PR TITLE
[BUG] writing datasets to file in ts format

### LINKED ISSUE #302: A function to save nested DataFrames to files
**Is your feature request related to a problem? Please describe.**
Constructing nested dataframes is time-consuming sometimes depending on the original format of the data.
 
**Describe the solution you'd like**
A function to save the nested dataframe to a file and another one to load it quickly. (the load function exists already probably)

### LINKED ISSUE #3499: [BUG]’write_dataframe_to_tsfile‘ doesn't work for length-unequal series
**Describe the bug**
I'm trying to make my own .ts format dataset using the function write_dataframe_to_tsfile.
There is no clear statement about whether length-unequal time series is  supported in the definition of function 'write_dataframe_to_tsfile'. 
There is a parameter named 'equal_lenth', which indicates whether each series is of equal length. And this parameter is 'False' in default. So I think it means this fuction support length-uneuqal series. 
 However, if I provide **a time series with 2 variates, 250 cases in different length**,  this function will returns '**ValueError: all input arrays must have the same shape'**.If I keep those cases in same length, it works. 
I think it maybe a bug? 

**To Reproduce**
I provide the following code to reproduce the bug.
PS: one of length-equal variable df1 has been commented out.

```python
import pandas as pd
from sktime.datasets import write_dataframe_to_tsfile
from sktime.datatypes._panel._convert import from_multi_index_to_nested

##df1=pd.DataFrame([[1,1,1,1],[1,2,2,2],[2,3,3,3],[2,4,4,4],[2,5,5,5],[3,6,6,6]],columns=['a','b','c','d'])
df1=pd.DataFrame ([[1,1,1,1],[1,2,2,2],[2,3,3,3],[2,4,4,4],[3,5,5,5],[3,6,6,6]],columns=['a','b','c','d'])

df1.set_index(keys=['a','b'],inplace=True)
print(df1)

Y_nested = from_multi_index_to_nested(df1)
##print(Y_nested.head())
write_dataframe_to_tsfile(data=Y_nested,equal_length= False,fold='_TRAIN',path='./path')
```

**Expected behavior**

**Additional context**
Providing length-unequal time series is important, because even the same motion may cost time more or less. 
Besides, there are some length-unequal time series archieve in UCR archieve, such like Basic_Motions so  i think sktime is supposed to provide a way for user to create a length-unequal time series in .ts format. 

**Versions**
sktime 0.13.2

</details>

### PR DESCRIPTION
**Pull request by TonyBagnall**
#### Reference Issues/PRs

fixes #3499

@achieveordie could you take a look please?

#### What does this implement/fix? Explain your changes.

this PR fixes a problem with writing data to .ts format and introduces a new method to make writing easier and more intuitive.
data_io contains two functions for saving ts files: write_dataframe_to_tsfile and write_ndarray_to_tsfile. write_dataframe_to_tsfile converts to ndarray and calls write_ndarray_to_tsfile. This does not work if the data frame contains unequal length series.  

In this PR write_dataframe_to_tsfile now gets the data metadata with check_is_scitype, then if equal length, converts to numpy, if not, iterates over the dataframe and writes in the format.

I have also implemented (and tested) a write_panel_to_tsfile. This gets data characteristics, if equal length converts to numpy then calls write_ndarray_to_tsfile, otherwise calls  write_dataframe_to_tsfile. The reason for this new function is there is no need for user to differentiate between dataframe and ndarray, and the default arguments to write_dataframe_to_tsfile and write_ndarray_to_tsfile are confusing (e.g. equal_length=False and  series_length=-1). There is no need to pass all these arguments in. 

There is some redundancy in write_panel_to_tsfile and write_dataframe_to_tsfile necessary to satisfy the deprecation policy: both functions call check_is_scitype and try to convert. I think write_dataframe_to_tsfile and write_ndarray_to_tsfile should be deprecated to private and the checks/conversion in write_dataframe_to_tsfile removed (so it simply writes from dataframe). I have not done this deprecation, but I can do so though.

#### Does your contribution introduce a new dependency? If yes, which one?
No
#### What should a reviewer concentrate their feedback on?
1. The testing writes and deletes files. Not sure if this is the preferred way of doing this
2. Should we deprecate write_ndarray_to_tsfile and write_dataframe_to_tsfile to being private? I think we should

##### For all contributions
- [x] I've added unit tests and made sure they pass locally.
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.

### CODE REVIEW

**Review by fkiraly — DISMISSED** [review]
@TonyBagnall, please do not change pytest configurations in your PR

**Comment by TonyBagnall** [comment]
> @TonyBagnall, please do not change pytest configurations in your PR

ah yes, sorry, I have reverted setup.cfg. I can't run pytests without removing those lines and forgot to put them back in. I think @chrisholder and @MatthewMiddlehurst have the same issue, we can look in detail and post an issue if its not just our set up causing the problem.

**Comment by TonyBagnall** [comment]
> @TonyBagnall, please do not change pytest configurations in your PR
@fkiraly I have reverted the inadvertent changes to config, could you please unblock this? thanks.

**Review by achieveordie — COMMENTED** [review]
Sorry, this had gone under my radar. I've added a few comments, everything else seems good. I have my reservations regarding the 3 nested loops, but I suppose there's no other way to do it.

File: sktime/datasets/_data_io.py [hunk 1]
```diff
@@ -1550,24 +1596,59 @@ def write_dataframe_to_tsfile(
...
-        missing_values=missing_values,
-        comment=comment,
-        fold=fold,
+        raise ValueError(f"Data provided must be a DataFrame, passed a {type(data)}")
+    data_valid, _, metadata = check_is_scitype(
+        data, scitype="Panel", return_metadata=True
     )
+    if not data_valid:
+        raise ValueError("DataFrame provided is not a valid type")
+    metadata
```
[the remark this answers is not in the document]
Comment (TonyBagnall) [reply sktime/datasets/_data_io.py:1605]: oops! Good spot

File: sktime/datasets/_data_io.py [hunk 2]
```diff
@@ -1493,6 +1499,46 @@ def write_tabular_transformation_to_arff(
...
+
+    """ Writes the header info for a ts file"""
+    file.write(f"@problemName {problem_name}\n")
+    file.write("@timestamps false\n")
+    file.write(f"@univariate {str(univariate).lower()}\n")
+    file.write(f"@equalLength {str(equal_length).lower()}\n")
+    if series_length > 0 and equal_length:
+        file.write(f"@seriesLength {series_length}\n")
+    # write class label line
+    if class_label is not None:
```
[the remark this answers is not in the document]
Comment (TonyBagnall) [reply sktime/datasets/_data_io.py:1533]: I agree,  but to do this we will need to update check_is_scitype to detect if the problem is a regression problem. I am not willing to dive into datatypes at the moment, so maybe we can leave this as a future issue?

[the remark this answers is not in the document]
Comment (achieveordie) [reply sktime/datasets/_data_io.py:1535]: I agree, we'll keep this in mind when introducing regression data compatibility in the near future.

**Review by achieveordie — APPROVED** [review]
LGTM

**Review by MatthewMiddlehurst — APPROVED** [review]

**Pull request merged**